rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,885 @@
+ * Raises RuntimeError if the underlying rcl library returns an error when sending the request.
+ *
+ * \param[in] pyaction_client The action client to use when sending the request.
+ * \param[in] pycancel_request The request message to send.
+ * \return sequence_number PyLong object representing the index of the sent request, or
+ * \return NULL if there is a failure.
+ */
+static PyObject *
+rclpy_action_send_cancel_request(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  SEND_SERVICE_REQUEST(cancel)
+}
+
+/// Take an action cancel response.
+/**
+ * Raises ValueError if pyaction_client does not have the correct capsule type.
+ * Raises RuntimeError if the underlying rcl library returns an error when taking the response.
+ *
+ * \param[in] pyaction_client The action client to use when sending the request.
+ * \param[in] pycancel_response_type An instance of the response message type to take.
+ * \return 2-tuple sequence number and received response, or
+ * \return None if there is no response, or
+ * \return NULL if there is a failure.
+ */
+static PyObject *
+rclpy_action_take_cancel_response(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  TAKE_SERVICE_RESPONSE(cancel)
+}
+
+#define TAKE_MESSAGE(Type) \
+  PyObject * pyaction_client; \
+  PyObject * pymsg_type; \
+  if (!PyArg_ParseTuple(args, "OO", & pyaction_client, & pymsg_type)) { \
+    return NULL; \
+  } \
+  rcl_action_client_t * action_client = (rcl_action_client_t *)PyCapsule_GetPointer( \
+    pyaction_client, "rcl_action_client_t"); \
+  if (!action_client) { \
+    return NULL; \
+  } \
+  PyObject * pymetaclass = PyObject_GetAttrString(pymsg_type, "__class__"); \
+  create_ros_message_signature * create_ros_message = get_capsule_pointer( \
+    pymetaclass, "_CREATE_ROS_MESSAGE"); \
+  assert(create_ros_message != NULL && \
+    "unable to retrieve create_ros_message function, type_support mustn't have been imported"); \
+  destroy_ros_message_signature * destroy_ros_message = get_capsule_pointer( \
+    pymetaclass, "_DESTROY_ROS_MESSAGE"); \
+  assert(destroy_ros_message != NULL && \
+    "unable to retrieve destroy_ros_message function, type_support mustn't have been imported"); \
+  void * taken_msg = create_ros_message(); \
+  if (!taken_msg) { \
+    Py_DECREF(pymetaclass); \
+    return PyErr_NoMemory(); \
+  } \
+  rcl_ret_t ret = rcl_action_take_ ## Type(action_client, taken_msg); \
+  if (ret != RCL_RET_OK) { \
+    if (ret != RCL_RET_ACTION_CLIENT_TAKE_FAILED) { \
+      /* if take failed, just do nothing */ \
+      destroy_ros_message(taken_msg); \
+      Py_DECREF(pymetaclass); \
+      Py_RETURN_NONE; \
+    } \
+    PyErr_Format(PyExc_RuntimeError, \
+      "Failed to take " #Type " with an action client: %s", rcl_get_error_string().str); \
+    rcl_reset_error(); \
+    destroy_ros_message(taken_msg); \
+    Py_DECREF(pymetaclass); \
+    return NULL; \
+  } \
+  convert_to_py_signature * convert_to_py = get_capsule_pointer(pymetaclass, "_CONVERT_TO_PY"); \
```

Comment:
Recommend getting all the capsule pointers at the top so `Py_DECREF(pymetaclass)` can be in fewer error blocks.